### ISSUE TITLE
ATECC608A TPM reading and seamless integration using PKCS11 library in AWS IOT device SDK

### ISSUE THREAD

**Issue by vishalSpintly** [aws/aws-iot-device-sdk-embedded-C#1772]
Hello,
We have an IoT device running Linux.
Currently, the certificates are stored on file system and use the AWS IOT Core C library to connect and authenticate with the AWS servers.

We are evaluating on our device, a ATECC608A TPM ( see: http://www.microchip.com/ATECC608A) connected over an I2C interface to store the keys.
So far, using resources available on the internet, we are able to configure the PKCS11 interface and able to communicate with the TPM device.

My area of exploration next is how do I use the open source PKCS11 infrastructure in conjunction with the AWS IOT Core ( C) library to connect and authenticate with the AWS servers seemlessly.

I am not able to find any guidance on the internet, and since I am new to TPM and related software technologies, i tend to get lost in my efforts.

Since this thread touches upon a use case similar to mine, I am posting my query in here.

I really need to figure out as to how I need to go about achieving what i need to get done in the light of a lack of meaningful resources and guides on the same,

Is this requirement supported? How do I enable the same to function on my linux board.

Please do guide. Any pointers are appreciated.

**Comment by pvyawaha** [comment]
Hello @vishalSpintly ,

You can refer to the PKCS11 layer for ESS608A here - https://github.com/aws/amazon-freertos/blob/main/libraries/abstractions/pkcs11/ecc608a/core_pkcs11_secure_element.c

And here is my example for key provisioning - https://github.com/aws/amazon-freertos/tree/main/demos/dev_mode_key_provisioning

**Comment by vishalSpintly** [comment]
My device uses Linux, and i have the PKCS11 infrastructure up and running on the OS.
However, Currently the code to connect and authenticate to AWS IOT Core ( mbedTLS library) uses credentials with a path pointing to a file that has the private key stored on the file system.
What can you recommend as an alternative way to configure the credentials by making use of PKCS#11?

**Comment by vishalSpintly** [comment]
Hello @pvyawaha ,

We are using mbedTLS to establish the connection with the AWS IOT server.
How do i go about using this implementation of the PKCS11 layer in my use case with mbedTLS. How do i configure mbedTLS to use the PKCS11 interface for reading the private key data from the TPM chip?
I am not able to identify any way to move forward and any guidance is appreciated.

**Comment by vishalSpintly** [comment]
Thank you for the pointers. these are helpful for understanding provisioning of the chip during development.
Since I am already using mbedTLS for authenticating with AWS IOT Service could you direct me to any effort related to the same?

**Comment by abhidixi11** [comment]
Hello @vishalSpintly does information above answer your questions ?

**Comment by vishalSpintly** [comment]
Hello @abhidixi11 ,

Unfortunately it does not. So the PKCS11 layer is not of any interest to me.

At https://github.com/MicrochipTech/mbedtls-examples is an integration of the mbedTLS library with the ATECC608A.

We use the mbedTLS library to establish the connection with the AWS IOT Core service.
With the above listed library integration it should be  possible to seamlessly connect to AWS IOT Core using credentials off the ATECC608A.

However the mbedTLS-examples repository that is archived on the said github repository seems to be broken, throwing errors in compilation due to missing functions ( functions being referenced, but not defined anywhere in the source files ).

see: https://github.com/MicrochipTech/mbedtls-examples/issues/10

So basically I am stuck without any way forward unless I get my hands on relevant and updated files.

regards,
Vishal

**Comment by johnrhen** [comment]
Hello @vishalSpintly,

You might find the [mbedtls_pkcs11_posix](https://github.com/johnrhen/aws-iot-device-sdk-embedded-C/blob/main/platform/posix/transport/include/mbedtls_pkcs11_posix.h) files to be helpful. They contain an implementation for the transport interface using a mutually authenticated TLS connection with MbedTLS for TLS and corePKCS11 for secure credential management. You can view a demonstration of MbedTLS and corePKCS11 working together to connect to IoT Core in the [Fleet Provisioning demo](https://github.com/johnrhen/aws-iot-device-sdk-embedded-C/tree/main/demos/fleet_provisioning/fleet_provisioning_with_csr). While our example uses the [corePKCS11 library](https://github.com/FreeRTOS/corePKCS11), you should be able to modify it to work with your configured PKCS11 infrastructure. Let us know if this information is helpful to you, and feel free to ask more questions.

Thank you,
John

**Comment by vishalSpintly** [comment]
Hello John,

This looks promising. 
Let me study this and get back to you with further questions.

regards,
Vishal

**Comment by vishalSpintly** [comment]
Hello John,

In our case, having configured the PKCS11 Library support in the cryptoauth library from Microchip, we go on to install the library (PKCS11 provider) at  "/usr/lib/libcryptoauth.so"
The PKCS11 slot to be used to access the private key & certificates onboard the ATECC608A would be for example "pkcs11:slot-id=1".

Considering that AWS corePKCS11 library implements the standard PKCS11 interfaces, is there any place where the AWS IOT Core library can be configured to use the said provider & slot in Code OR configuration file. I know IOT Greengrass allows you to do this.

Or, is there a need to replace the corePKCS11 Library and associated function calls within AWS IoT Core, with the corresponding references into cryptoauth PKCS11 library - which seems unnecessary & circuitous in my understanding.
 
Let me know you thoughts. 
Microchip seems to have been already accomplished this integration at their end - based on the buggy resources i have listed in my previous comments.

**Comment by paulbartell** [comment]
@vishalSpintly: We have had a few requests for a similar feature in the past. The simplest way to accomplish this is to use a combination of cryptoauthlib built as a pkcs11 module and openssl with [libp11](https://github.com/OpenSC/libp11).

You will first need to decide if you plan on using [_p11-kit_](https://p11-glue.github.io/p11-glue/p11-kit/manual/) to proxy access to your pkcs11 module. p11-kit allows concurrent access to a pkcs11 module(libcryptoauth.so in your case) and some other common linux desktop oriented features. If concurrent access is not necessary, you can use _libp11_ to access your pkcs11 module (libcryptoauth.so) directly when initializing a tls connection with openssl.

In addition to the documentation you [linked to earlier](https://github.com/MicrochipTech/cryptoauthlib/wiki/PKCS11-Linux-Setup), I would like to call your attention to this page [with a bit more openssl specific info.](https://github.com/MicrochipTech/cryptoauthlib/wiki/OpenSSL-Reference).

When building the aws-iot c-sdk, make sure to use openssl as your tls library. I've staged some changes that add pkcs11 support to openssl_posix.c [in a branch on my fork over here](https://github.com/paulbartell/aws-iot-device-sdk-embedded-C/commit/a0b419cb726e99a5786ecd54ac0c1cea1c1db225). I have not had a chance to test it with an ECC608A or SoftHSM2 yet, but I plan on doing so sometime next week.

Note: The ATECC608 does not implement the [TCG](https://trustedcomputinggroup.org/) TPM api, so using the term TPM to describe could be a bit confusing. There is an open source project that implements a TPM->pkcs11 shim which could be useful as a reference: [tpm2-pkcs11](https://github.com/tpm2-software/tpm2-pkcs11).

**Comment by vishalSpintly** [comment]
@paulbartell 

Thanks !!

The functionality being targeted will run on a IOT gateway device ( which delicately brokers the message exchange between the AWS cloud and the individual IOT devices). 
I believe libp11 should suffice as there is no concurrent access required.

>When building the aws-iot c-sdk, make sure to use openssl as your tls library. I've staged some changes that add pkcs11 >support to openssl_posix.c in a branch on my fork over here. I have not had a chance to test it with an ECC608A or SoftHSM2 >yet, but I plan on doing so sometime next week.

I will check the above on my end. Hopefully i get the configurations right.
Do let me know you test findings. I will be keen to know the results.

>Note: The ATECC608 does not implement the TCG TPM api, so using the term TPM to describe could be a bit confusing. Noted and updated.

**Comment by vishalSpintly** [comment]
> When building the aws-iot c-sdk, make sure to use openssl as your tls library. 

i have your repository cloned. 
Can you tell me where & how I need to specify openssl as tls library during the build of aws-iot-c-sdk ?

**Comment by abhidixi11** [comment]
Hello @vishalSpintly we have the instruction in Readme about OpenSSL version supported Linux platforms:
https://github.com/aws/aws-iot-device-sdk-embedded-C#build-dependencies
https://github.com/aws/aws-iot-device-sdk-embedded-C#building-and-running-demos
Thanks.

**Comment by vishalSpintly** [comment]
@abhidixi11 If your comment is in response to my query "where & how I need to specify openssl as tls library during the build of aws-iot-c-sdk", I have not received the answer yet.

For example, there could be some flag that needs to be set during cmake on the aws-iot-c-sdk build to indicate the preference for openssl over other options avaialble. If yes, what is that flag.

**Comment by abhidixi11** [comment]
@vishalSpintly 
At this time, this repository has the code which works with openSSL for TLS connection.  You can replace transport interface yourself if you want to make use of any other TLS library.

**Comment by vishalSpintly** [comment]
@abhidixi11 
I am not particular about the TLS library. 

My clear intent is to get the aws-iot-c-sdk based connection to AWS service using the ATECC608A seamlessly. 
If this is done using openssl, or any other library there will be nor reason to complain as long as it works by using the credentials off the ATECC608A.

I can only try your suggestions at this moment.
So, I will give this [fork](https://github.com/paulbartell/aws-iot-device-sdk-embedded-C/commit/a0b419cb726e99a5786ecd54ac0c1cea1c1db225) a fresh test in the hope that it does what i have set out to achieve.

**Comment by paulbartell** [comment]
@vishalSpintly : Regarding the marketing literature: [amazon-freertos](https://github.com/aws/amazon-freertos) does indeed support the ATECC608. That implementation is targeted to a bare-metal / FreeRTOS environment and is not particularly well suited for use in a linux environment. It's certainly possible to use that implementation on linux, but it is not consistent with how pkcs11 is typically used in a linux environment.

Re: Selecting a TLS library. It's expected that you may use a different build system to compile the portions of the sdk that are necessary for your application. This involves selecting which libraries and transport implementation that is best for your particular application. Openssl is the default option when using the included cmake build files.

**Comment by paulbartell** [comment]
@vishalSpintly : I've [updated the feature branch](https://github.com/paulbartell/aws-iot-device-sdk-embedded-C/commits/paulbartell/openssl_libp11_support) with some bug fixes and documentation. Please let me know if you have any questions about the setup.

**Comment by vishalSpintly** [comment]
Hi @paulbartell I could not check on this as i had to focus on another critical task that came up. 
I am now back to working on this. 
My plan is to start afresh, by generating a fresh test setup with the Linux OS and support tools on our development environment first. 
I will document my steps and post my queries on this thread in my subsequent posts.

Look forward to your support.

**Comment by lmajewski** [comment]
@paulbartell @vishalSpintly - is there any plan to upstream patches from https://github.com/paulbartell/aws-iot-device-sdk-embedded-C/commits/paulbartell/openssl_libp11_support to https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/main/platform/posix/transport ?

It looks like the only option for ATECC608 (or any other HSM) is to try to follow mbedtls_pkcs11_posix.c ?
This is a bit strange, as openssl with pkcs11 and cryptoautlhib works on Linux without issues.
